File: Anya_Seth/Sliding Window/Minimum Window Substring.java
class Solution {
    public String minWindow(String s, String t) {
        int[] freqT = new int[128];
        int[] freqS = new int[128];

        int start = 0;
        int end = 0;
        int count = 0;
        int startIndx = -1;
        int endIndx = -1;
        int minLength = s.length() + 1;
    
        for(int i = 0 ; i < t.length() ; i++){
            char ch = t.charAt(i);
            if(freqT[ch] == 0){
                count++;
            }
            freqT[ch]++;
        }
        while(end < s.length()){
            char c = s.charAt(end);
            freqS[c]++;
            if(freqS[c] == freqT[c]){
                count--;
            }
            while(count == 0){
                if(end - start + 1 < minLength){
                    startIndx = start;
                    endIndx = end;
                    minLength = end - start + 1;
                }
            char ch = s.charAt(start);
            if(freqS[ch] == freqT[ch]){
                count++;
            }
            start++;
            freqS[ch]--;
        }
        end++;
        }
          return startIndx == -1 ? "" : s.substring(startIndx, endIndx + 1);
    }
}

File: Lakshay Garg/Leetcode/Sliding Window/Minimum Window Substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int start=0,end=0,mini=INT_MAX,f=-1,l=-1,count=0;
        vector<int>ft(126,0);
        vector<int>fs(126,0);
        for(char c:t){
            ft[c]++;
        }
        for(int i=0;i<126;i++){
            if(ft[i]!=0) count++;
        }
        while(end<s.size()){
            char c=s[end];
            fs[c]++;
            if(fs[c]==ft[c]) count--;
            while(count==0){
                if(mini>end-start+1){
                    f=start,l=end,mini=end-start+1;
                }
                char ch=s[start];
                if(fs[ch]==ft[ch]) count++;
                fs[ch]--;
                start++;
            }
            end++;
        }
        if(f==-1) return "";
        return s.substr(f,mini);
    }
};
